### PR TITLE
paddlebox about box_wrapper

### PR DESCRIPTION
**Pull request by hutuxian**
Refine PaddleBox Framework：
* Add MetricMsg util class, which can calculate metrics like AUC, bucket_error, COPC.
* Replace FeedPass with new interface: BeginFeedPass & EndFeedPass
* Refactor Pull/Push Sparse Function in box_wrapper.
* Use CUDA Kernel to copy keys and copy feasign between tensor and boxps struct.
* Cache copied keys in pull sparse in order to reuse it in push period.

### CODE REVIEW

**Comment by gavin1332** [comment]
you should describe what this pr do in your first comment.

File: paddle/fluid/framework/fleet/box_wrapper.h [hunk 1]
```diff
@@ -14,27 +14,117 @@ limitations under the License. */
...
+  double _local_abserr = 0;
+  double _local_sqrerr = 0;
+  double _local_pred = 0;
+  double _auc = 0;
+  double _mae = 0;
+  double _rmse = 0;
+  double _actual_ctr = 0;
+  double _predicted_ctr = 0;
+  double _size;
+  double _bucket_error = 0;
```
Comment (gavin1332) [inline paddle/fluid/framework/fleet/box_wrapper.h:103]: in order to distinguish the join-calculator and the update-calculator, it is suggested that these statistics stored in a LoDTensor carried by corresponding scope.

File: paddle/fluid/framework/fleet/box_wrapper.h [hunk 2]
```diff
@@ -46,44 +136,185 @@ class BoxWrapper {
...
+
+  void Finalize() {
+    VLOG(3) << "Begin Finalize";
+    if (nullptr != s_instance_) {
+      s_instance_->boxps_ptr_->Finalize();
+    }
+  }
+
+  void SaveBase(const char* batch_model_path, const char* xbox_model_path,
+                boxps::SaveModelStat& stat) {  // NOLINT
```
Comment (gavin1332) [inline paddle/fluid/framework/fleet/box_wrapper.h:192]: it is not good to expose `boxps` namespace out of box_wrapper, it is better to wrap `boxps`-related data in a paddle-friendly data structure.

File: paddle/fluid/framework/fleet/box_wrapper.h [hunk 3]
```diff
@@ -46,44 +136,185 @@ class BoxWrapper {
...
       }
     }
     return s_instance_;
   }
 
+  const std::unordered_set<std::string>& GetOmitedSlot() const {
+    return slot_name_omited_in_feedpass_;
+  }
+
+  struct MetricMsg {
```
Comment (gavin1332) [inline paddle/fluid/framework/fleet/box_wrapper.h:226]: could it designed to be an element of the scope as LoDTensor be?

File: paddle/fluid/framework/fleet/box_wrapper.h [hunk 4]
```diff
@@ -46,44 +136,185 @@ class BoxWrapper {
...
+
+   private:
+    std::string label_varname_;
+    std::string pred_varname_;
+    int is_join_;
+    BasicAucCalculator* calculator;
+  };
+
+  int PassFlag() const { return pass_flag_; }
+  void FlipPassFlag() { pass_flag_ = 1 - pass_flag_; }
```
Comment (gavin1332) [inline paddle/fluid/framework/fleet/box_wrapper.h:250]: if auc statistics stored in a specific scope, we do not need to flip or switch flag any longer

**Review by gavin1332 — COMMENTED** [review]
request changes

**Review by gavin1332 — APPROVED** [review]
LGTM

File: paddle/fluid/framework/fleet/box_wrapper.h (hunk 1 above)
Comment (hutuxian) [reply paddle/fluid/framework/fleet/box_wrapper.h:103]: Thanks, will make it a future work.

File: paddle/fluid/framework/fleet/box_wrapper.h (hunk 2 above)
Comment (hutuxian) [reply paddle/fluid/framework/fleet/box_wrapper.h:192]: Thanks, will make it a future work

**Pull request merged**